server/src/test/java/org/opensearch/indices/IndicesRequestCacheTests.java

```diff
@@ -699,47 +691,179 @@ public void testCacheCleanupBasedOnStaleThreshold_StalenessLesserThanThreshold()
+        ShardId shardId = indexShard.shardId();
+        ThreadPool threadPool = getThreadPool();
+        Settings settings = Settings.builder().put(INDICES_REQUEST_CACHE_STALENESS_THRESHOLD_SETTING.getKey(), "0.51").build();
+        IndicesRequestCache cache = new IndicesRequestCache(settings, (currentShardId -> {
+            IndexService indexService = null;
+            try {
+                indexService = indicesService.indexServiceSafe(shardId.getIndex());
+            } catch (IndexNotFoundException ex) {
+                return Optional.empty();
+            }
+            return Optional.of(new IndicesService.IndexShardCacheEntity(indexService.getShard(shardId.id())));
+        }), new CacheModule(new ArrayList<>(), settings).getCacheService(), threadPool);
+        Directory dir = newDirectory();
+        IndexWriter writer = new IndexWriter(dir, newIndexWriterConfig());
+        writer.addDocument(newDoc(0, "foo"));
+
+        DirectoryReader reader = OpenSearchDirectoryReader.wrap(DirectoryReader.open(writer), new ShardId("foo", "bar", 1));
+        DirectoryReader secondReader = OpenSearchDirectoryReader.wrap(DirectoryReader.open(writer), new ShardId("foo", "bar", 1));
+        TermQueryBuilder termQuery = new TermQueryBuilder("id", "0");
+        BytesReference termBytes = XContentHelper.toXContent(termQuery, MediaTypeRegistry.JSON, false);
+
+        // Get 2 entries into the cache from 2 different readers
+        IndicesService.IndexShardCacheEntity entity = new IndicesService.IndexShardCacheEntity(indexShard);
+        Loader loader = new Loader(reader, 0);
+        cache.getOrCompute(entity, loader, reader, termBytes);
+        assertEquals(1, cache.count());
+        // test the mappings
+        String readerCacheKeyId = getReaderCacheKeyId(reader);
+        ConcurrentMap<ShardId, HashMap<String, Integer>> cleanupKeyToCountMap = cache.cacheCleanupManager.getCleanupKeyToCountMap();
+        assertEquals(1, (int) cleanupKeyToCountMap.get(shardId).get(readerCacheKeyId));
+
+        IndicesService.IndexShardCacheEntity secondEntity = new IndicesService.IndexShardCacheEntity(indexShard);
+        loader = new Loader(secondReader, 0);
+        cache.getOrCompute(secondEntity, loader, secondReader, termBytes);
+        // test the mapping
+        readerCacheKeyId = getReaderCacheKeyId(secondReader);
+        assertEquals(2, cache.count());
+        assertEquals(1, (int) cleanupKeyToCountMap.get(shardId).get(readerCacheKeyId));
+
+        // Close the reader, to create stale entries
+        reader.close();
+        // 1 out of 2 keys ie 50% are now stale.
+        assertEquals(1, cache.cacheCleanupManager.getStaleKeysCount().get());
+        // cache count should not be affected
+        assertEquals(2, cache.count());
+
+        // test the mapping
+        readerCacheKeyId = getReaderCacheKeyId(reader);
+        assertEquals(1, (int) cleanupKeyToCountMap.get(shardId).get(readerCacheKeyId));
```

Comment:
why `cleanupKeyToCountMap` still has the entry for reader which is closed ? The entry is accounted in staleKeys so it should be removed from the map